### DISCUSSION TITLE
Equivalent of `state_diffs` in ethereum for aptos?

### DISCUSSION THREAD

**Discussion by 0xnullifier** [aptos-labs/aptos-developer-discussions#430] [Questions]
### Discord user ID

utkarshdagoat

### Describe your question in detail.

Is there a concept of "state diffs" like in ethereum which allow for data manupulation with `eth_call` method for aptos? I am trying to implement a rollup on aptos and have a hard time figuring out on how to keep aptos L1 as my Data availability layer is this something possible with aptos as I cannot see any rpc method allowing for direct state manipulations in the api documentation?

### Which operating system are you using?

Linux (Ubuntu, Fedora, Windows WSL, etc.)

### Which SDK or tool are you using? (if any)

Rust SDK

**Comment by MoonShiesty** [comment]
a "change set" is included in transaction receipts on aptos, which includes all the state changes from the transaction

there's no rpc subscription api right now. to create a production rollup system you should look into the grpc/indexer api which can stream and filter all the transactions along with the change set

you can find the change set api documented in the get/waitTransactions api. the type of transactions you need to look at are `user_transactions` the child attribute is `changes`


https://aptos.dev/en/build/apis/fullnode-rest-api-reference#tag/transactions/POST/transactions

**Reply by 0xnullifier** [reply]
Thanks for the answer. With the Indexer api I should be able to construct the rollup state from aptos.

Is it also possible to push these `changes` in the `user_transactions`  only instead of the whole transaction?
Pushing only the `changes` allow for me to have more cost efficient because I can omit signatures when multiple transactions modify the same state. 

If I cannot send the `changes` can you suggest a way of pushing transactions onto aptos I am assuming `batch_send` is the make the most sense  for pushing a block of transaction from my rollup to aptos L1?

**Comment by MoonShiesty** [comment]
> Pushing only the changes allow for me to have more cost efficient because I can omit signatures when multiple transactions modify the same state.

can you explain in more detail what you mean

**Reply by 0xnullifier** [reply]
So basically for a block of transactions rather than submitting the original transaction data. I can submit the output of the transactions as `state_diffs` 
So if there are **multiple transactions in a block changing the same state** I can just report the difference in the final state and initial state and hence I can decrease the `transaction` data sent to the L1 and thereby being more cost efficient . This is a well known concept in validity  based ethereum rollups like `zkSync`.

If you would like to read more about Data availability check out this  [artcle](https://scalability.guide/posts/txdata_vs_statediffs/) by scalability guide


Well A bigger issue for me right now is generating proofs for execution of move bytecode. This is generally done by writing circuits for each operation in low level bytecode. So I don't know if I will continue with this or not?

**Reply by MoonShiesty** [reply]
> So if there are multiple transactions in a block changing the same state I can just report the difference in the final state and initial state and hence I can decrease the transaction data sent to the L1 and thereby being more cost efficient . This is a well known concept in validity based ethereum rollups like zkSync.

aptos has blocks but seems like you might have to parse each transaction's state diffs to generate a block's state diff

here's the block api: https://aptos.dev/en/build/apis/fullnode-rest-api-reference#tag/blocks/GET/blocks/by_height/{block_height}

**Reply by 0xnullifier** [reply]
Thank you but I have put this task on hold for now. I will just take whole transaction data.

On second hand if there is a move expert out there who can help me write circuits for the execution trace reach out to me on disc!

**Reply by MoonShiesty** [reply]
im not the expert but i know aptos used zk for aptos connect (keyless wallet)

here's the circuit: https://github.com/aptos-labs/aptos-core/tree/main/keyless/circuit

[this dev](https://x.com/mstrakastrak/status/1765832363282448685) might be good to talk to
